specifyweb/specify/migrations/0002_geo.py

```diff
@@ -0,0 +1,168 @@
+                ('modifiedbyagent', models.ForeignKey(db_column='ModifiedByAgentID', null=True, on_delete=specifyweb.specify.models.protect_with_blockers, related_name='+', to='specify.agent')),
+            ],
+            options={
+                'db_table': 'collectionobjectgroup',
+                'ordering': (),
+            },
+        ),
+        migrations.CreateModel(
+            name='CollectionObjectGroupJoin', # add as dependant to collection object group
+            fields=[
+                ('id', models.AutoField(db_column='collectionobjectgroupjoinid', primary_key=True, serialize=False)),
+                ('isprimary', models.BooleanField(blank=True, db_column='IsPrimary', null=True)),
+                ('issubstrate', models.BooleanField(blank=True, db_column='IsSubstrate', null=True)),
+                ('precedence', models.SmallIntegerField(blank=True, db_column='Precedence', null=True)),
+                ('version', models.IntegerField(blank=True, db_column='Version', default=0, null=True)),
+                ('timestampcreated', models.DateTimeField(db_column='TimestampCreated', default=django.utils.timezone.now)),
+                ('timestampmodified', models.DateTimeField(blank=True, db_column='TimestampModified', default=django.utils.timezone.now, null=True)),
+                ('text1', models.TextField(blank=True, db_column='Text1', null=True)),
+                ('text2', models.TextField(blank=True, db_column='Text2', null=True)),
+                ('text3', models.TextField(blank=True, db_column='Text3', null=True)),
+                ('integer1', models.IntegerField(blank=True, db_column='Integer1', null=True)),
+                ('integer2', models.IntegerField(blank=True, db_column='Integer2', null=True)),
+                ('integer3', models.IntegerField(blank=True, db_column='Integer3', null=True)),
+                ('boolean1', models.BooleanField(blank=True, db_column='Boolean1', null=True)),
+                ('boolean2', models.BooleanField(blank=True, db_column='Boolean2', null=True)),
+                ('boolean3', models.BooleanField(blank=True, db_column='Boolean3', null=True)),
+                ('parentcog', models.ForeignKey(db_column='ParentCOGID', on_delete=django.db.models.deletion.CASCADE, related_name='parentcojos', to='specify.collectionobjectgroup')),
+                ('childcog', models.OneToOneField(db_column='ChildCOGID', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='cojo', to='specify.collectionobjectgroup')),
+                ('childco', models.OneToOneField(db_column='ChildCOID', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='cojo', to='specify.collectionobject')),
+            ],
+            options={
+                'db_table': 'collectionobjectgroupjoin',
+                'ordering': (),
+            },
+        ),
+        migrations.AlterUniqueTogether(
```

Comment:
We still need uniqueness constraints for `IsPrimary` and `IsSubstrate` as these should only appear _once_ within each COG

Reply:
This is enforced in the business logic here.  It states that in the CoJo table, each parentcog can only have one `IsPrimary`/`IsSubstrate` child.  Does this look good, or want it changed?
https://github.com/specify/specify7/blob/eeafd17ff7eb64c036cbb354b4c45dfeecf3f848/specifyweb/businessrules/rules/cojo_rules.py#L4-L19

Reply:
I could do something like this to have a constraint on the database side
```
migrations.RunSQL(
            """
            ALTER TABLE collectionobjectgroupjoin
            ADD CONSTRAINT chk_single_primary_parentcog
            CHECK (
                (isprimary = TRUE AND
                (SELECT COUNT(*) 
                FROM collectionobjectgroupjoin AS cog2 
                WHERE cog2.parentcog = collectionobjectgroupjoin.parentcog 
                    AND cog2.isprimary = TRUE) <= 1)
                OR isprimary = FALSE
            );
            """,
            reverse_sql=''
        ),
```
but having it in the business logic works, as long as no one is just making manual edits to their database.

Reply:
Looks good! nobody should be manually editing the database...